### PR TITLE
chore: field editing bloc refactor and add mobile field editor

### PR DESCRIPTION
**Pull request by richardshiue**
- Reliance on field controller stems from requiring FieldInfo and not simply the FieldPB. This will no longer be necessary after field info refactor https://github.com/AppFlowy-IO/AppFlowy/pull/3633

- Need to wait 50 milliseconds for the field controller's fields to be correct in tests. this will also no longer be necessary after the field info refactor

- GridFieldActionSheetBloc has been merged into FieldEditorBloc. Now the latter serves both menus (shown in the screenshots below)

- The edit property page will come in the next PR. This one is too big already.

1. Adjust the field edit menu on grid, row detail page and left-side setting toolbar dropdown.

   image

   image

2. Add a field editing bottom sheet

   Screenshot 2023-11-21 at 11 15 20 PM

TODO:

- [x] fix field delete on desktop
- [x] make the mobile bottom sheet work

### Feature Preview

---

#### PR Checklist

- [x] My code adheres to [AppFlowy's Conventions](https://docs.appflowy.io/docs/documentation/software-contributions/conventions)
- [ ] I've listed at least one issue that this PR fixes in the description above.
- [ ] I've added a test(s) to validate changes in this PR, or this PR only contains semantic changes.
- [ ] All existing tests are passing.

### CODE REVIEW

File: frontend/appflowy_flutter/lib/mobile/presentation/bottom_sheet/bottom_sheet_database_field_editor.dart [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+  Widget _buildBody() {
+    return switch (viewMode) {
+      MobileDBBottomSheetViewMode.general => MobileDBFieldBottomSheetBody(
+          onAction: (action) {
+            switch (action) {
+              case MobileDBBottomSheetGeneralAction.typeOption:
+                debugPrint("gimme the type option editor!");
+                // setState(() {
+                //   viewMode = MobileDBBottomSheetViewMode.typeOption;
+                // });
```
Comment (LucasXu0) [inline frontend/appflowy_flutter/lib/mobile/presentation/bottom_sheet/bottom_sheet_database_field_editor.dart:106]: remove the unused code and debugprint.

File: frontend/appflowy_flutter/lib/mobile/presentation/bottom_sheet/bottom_sheet_database_field_editor.dart [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+      //       case MobileDBBottomSheetGeneralAction.typeOption:
+      //         setState(() {
+      //           viewMode = MobileDBBottomSheetViewMode.typeOption;
+      //         });
+      //         break;
+      //       default:
+      //         break;
+      //     }
+      //   },
+      // ),
```
Comment (LucasXu0) [inline frontend/appflowy_flutter/lib/mobile/presentation/bottom_sheet/bottom_sheet_database_field_editor.dart:139]: remove

File: frontend/appflowy_flutter/lib/mobile/presentation/bottom_sheet/bottom_sheet_database_field_editor.dart [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+              child: BottomSheetActionWidget(
+                svg: FlowySvgs.delete_s,
+                text: LocaleKeys.grid_field_delete.tr(),
+                onTap: () {
+                  onAction(MobileDBBottomSheetGeneralAction.delete);
+                },
+              ),
+            ),
+            const HSpace(8),
+            const Expanded(child: SizedBox.shrink()),
```
Comment (LucasXu0) [inline frontend/appflowy_flutter/lib/mobile/presentation/bottom_sheet/bottom_sheet_database_field_editor.dart:222]: unused code?

File: frontend/appflowy_flutter/lib/mobile/presentation/bottom_sheet/bottom_sheet_database_field_header.dart [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+              child: Icon(
+                Icons.arrow_back_ios_new_rounded,
+                size: 24.0,
+              ),
+            ),
+          ),
+        // field name
+        Expanded(
+          child: Text(
+            "Edit Field",
```
Comment (LucasXu0) [inline frontend/appflowy_flutter/lib/mobile/presentation/bottom_sheet/bottom_sheet_database_field_header.dart:34]: i18n

File: frontend/appflowy_flutter/lib/mobile/presentation/database/card/card_detail/widgets/mobile_create_row_field_button.dart [hunk 5]
```diff
@@ -21,21 +27,22 @@ class MobileCreateRowFieldButton extends StatelessWidget {
...
-              MobileCreateRowFieldScreen.routeName,
-              extra: {
-                MobileCreateRowFieldScreen.argViewId: viewId,
-                MobileCreateRowFieldScreen.argTypeOption: typeOption,
-              },
-            );
-          },
-          (r) => Log.error("Failed to create field type option: $r"),
+      onPressed: () {
+        TypeOptionBackendService.createFieldTypeOption(viewId: viewId).then(
```
Comment (LucasXu0) [inline frontend/appflowy_flutter/lib/mobile/presentation/database/card/card_detail/widgets/mobile_create_row_field_button.dart:31]: I suggest keeping `await` here to reduce the indent.

File: frontend/appflowy_flutter/lib/mobile/presentation/database/card/card_detail/mobile_card_detail_screen.dart [hunk 6]
```diff
@@ -24,12 +25,16 @@ class MobileCardDetailScreen extends StatefulWidget {
   const MobileCardDetailScreen({
     super.key,
     required this.rowController,
+    required this.fieldController,
   });
 
   static const routeName = '/MobileCardDetailScreen';
   static const argRowController = 'rowController';
   static const argCellBuilder = 'cellBuilder';
+  static const argFieldController = 'fieldController';
```
Comment (LucasXu0) [inline frontend/appflowy_flutter/lib/mobile/presentation/database/card/card_detail/mobile_card_detail_screen.dart:34]: is it the required parameter for every card detail page?

File: frontend/appflowy_flutter/lib/plugins/database_view/grid/presentation/widgets/header/field_type_option_editor.dart [hunk 7]
```diff
@@ -74,22 +73,36 @@ class FieldTypeOptionEditor extends StatelessWidget {
...
-    Key? key,
-  }) : super(key: key);
+  });
+
+  @override
+  State<SwitchFieldButton> createState() => _SwitchFieldButtonState();
+}
+
+class _SwitchFieldButtonState extends State<SwitchFieldButton> {
+  late final PopoverController _popoverController;
```
Comment (LucasXu0) [inline frontend/appflowy_flutter/lib/plugins/database_view/grid/presentation/widgets/header/field_type_option_editor.dart:88]: no need to use late

File: frontend/appflowy_flutter/lib/mobile/presentation/bottom_sheet/bottom_sheet_database_field_editor.dart (hunk 3 above)
Comment (richardshiue) [reply frontend/appflowy_flutter/lib/mobile/presentation/bottom_sheet/bottom_sheet_database_field_editor.dart:222]: No, it's used to make the width of the delete button line up correctly

File: frontend/appflowy_flutter/lib/mobile/presentation/database/card/card_detail/mobile_card_detail_screen.dart (hunk 6 above)
Comment (richardshiue) [reply frontend/appflowy_flutter/lib/mobile/presentation/database/card/card_detail/mobile_card_detail_screen.dart:34]: Yes it's needed for the `FieldEditorBloc`

File: frontend/appflowy_flutter/lib/mobile/presentation/bottom_sheet/bottom_sheet_database_field_editor.dart (hunk 3 above)
Comment (Saad2344) [reply frontend/appflowy_flutter/lib/mobile/presentation/bottom_sheet/bottom_sheet_database_field_editor.dart:205]: - [شي أن ok ht]()

**Pull request merged**